engine/api/v2_workflow_run_craft.go

```diff
@@ -0,0 +1,469 @@
+						log.Error(ctx, "V2WorkflowRunCraftChan> error on workflow run %s: %v", id, err)
+					}
+				},
+			)
+		case <-ticker.C:
+			ids, err := workflow_v2.LoadCratingWorkflowRunIDs(api.mustDB())
+			if err != nil {
+				log.Error(ctx, "V2WorkflowRunCraftChan> unable to start tx: %v", err)
+				continue
+			}
+			for _, id := range ids {
+				api.GoRoutines.Exec(
+					ctx,
+					"V2WorkflowRunCraftDB-"+id,
+					func(ctx context.Context) {
+						ctx = telemetry.New(ctx, api, "api.V2WorkflowRunCraftDB", nil, trace.SpanKindUnspecified)
+						if err := api.craftWorkflowRunV2(ctx, id); err != nil {
+							log.Error(ctx, "V2WorkflowRunCraftDB> error on workflow run %s: %v", id, err)
+						}
+					},
+				)
+			}
+		}
+	}
+}
+
+func (api *API) craftWorkflowRunV2(ctx context.Context, id string) error {
+	_, next := telemetry.Span(ctx, "api.craftWorkflowRunV2.lock")
+	lockKey := cache.Key("api:craftWorkflowRunV2", id)
+	b, err := api.Cache.Lock(lockKey, 5*time.Minute, 0, 1)
+	if err != nil {
+		next()
+		return err
+	}
+	if !b {
+		log.Debug(ctx, "api.craftWorkflowRunV2> run %d is locked in cache", id)
+		next()
+		return nil
+	}
+	next()
+	defer func() {
+		_ = api.Cache.Unlock(lockKey)
+	}()
+
+	_, next = telemetry.Span(ctx, "api.craftWorkflowRunV2.LoadRunByID")
+	run, err := workflow_v2.LoadRunByID(ctx, api.mustDB(), id)
+	if sdk.ErrorIs(err, sdk.ErrNotFound) {
+		next()
+		return nil
+	}
+	if err != nil {
+		next()
+		return sdk.WrapError(err, "unable to load workflow run %s", id)
+	}
+	next()
+
+	if run.Status != sdk.StatusWorkflowRunCrafting {
+		return nil
+	}
+
+	vcsServer, err := vcs.LoadVCSByID(ctx, api.mustDB(), run.ProjectKey, run.VCSServerID)
+	if err != nil {
+		return err
+	}
+	repo, err := repository.LoadRepositoryByVCSAndID(ctx, api.mustDB(), vcsServer.ID, run.RepositoryID)
+	if err != nil {
+		return err
+	}
+
+	u, err := user.LoadByID(ctx, api.mustDB(), run.UserID)
+	if err != nil {
+		return err
+	}
+
+	// Build run context
+	runContext := buildRunContext(*run, *vcsServer, *repo, *u)
+	run.Contexts = runContext
+
+	wref := WorkflowRunEntityFinder{
+		run:                    *run,
+		runRepo:                *repo,
+		runVcsServer:           *vcsServer,
+		actionsCache:           make(map[string]sdk.V2Action),
+		workerModelCache:       make(map[string]sdk.V2WorkerModel),
+		repoCache:              make(map[string]sdk.ProjectRepository),
+		vcsServerCache:         make(map[string]sdk.VCSProject),
+		repoDefaultBranchCache: make(map[string]string),
+		userName:               u.Username,
+	}
+
+	// Retrieve all deps
+	for jobID := range run.WorkflowData.Workflow.Jobs {
+		j := run.WorkflowData.Workflow.Jobs[jobID]
+
+		// Get worker model
+		completeName, msg, err := wref.searchEntity(ctx, api.mustDB(), api.Cache, j.WorkerModel, sdk.EntityTypeWorkerModel)
+		if err != nil {
+			log.ErrorWithStackTrace(ctx, err)
+			return stopRun(ctx, api.mustDB(), run, &sdk.V2WorkflowRunInfo{
```

Comment:
why don't we use the same database transaction in the "stop" function ?

Reply:
We are not in a transaction